// filmsurf-routes/lib/persistence.js
var mongoose = require('mongoose');

console.log('Opening mongo connection to ' + process.env.MONGOLAB_URI);

mongoose.connect(process.env.MONGOLAB_URI);

mongoose.connection.addListener('open', function() {
    console.log('Mongo connection opened');
});

var Schema = mongoose.Schema;

var CastMember = new Schema({ id: String, name: String, characters: [String] });

var Director = new Schema({ name: String });

var Links = {
    self: String,
    alternate: String,
    cast: String,
    clips: String,
    reviews: String,
    similar: String
};

var Movie = new Schema({
    _id: { type: String, index: { unique: true } },
    title: String,
    year: Number,
    genres: [String],
    mpaa_rating: String,
    runtime: Number,
    critics_consensus: String,
    release_dates: { theater: String, dvd: String },
    ratings: {
        critics_rating: String,
        critics_score: Number,
        audience_rating: String,
        audience_score: Number
    },
    synopsis: String,
    posters: {
        thumbnail: String,
        profile: String,
        detailed: String,
        original: String
    },
    abridged_cast: [CastMember],
    abridged_directors: [Director],
    studio: String,
    alternate_ids: { imdb: String },
    links: Links
});

var SearchResult = new Schema({
    _id: { type: String, index: { unique: true } },
    total: Number,
    movies: [Movie],
    links: Links,
    link_template: String
});

var MovieModel = mongoose.model('Movie', Movie);
var SearchResultModel = mongoose.model('SearchResult', SearchResult);

exports.persistMovie = function (id, movie) {
    console.log('Persisting movie: ' + JSON.stringify(movie));
};

exports.persistSearch = function (query, results) {
    console.log('Persisting search: ' + JSON.stringify(results));
};
